src/index.tsx

```diff
@@ -32,15 +34,19 @@ import './index.css';
 function Index() {
   const user = useSelector((state: RootState) => state.user.currentUser);
 
-  const loaderWrapper = async (fn: () => Promise<TypePromise<any>>) => {
+  const loaderWrapper = async (
+    fn: () => Promise<TypePromise<any>>,
+    allowedUsers: 'student' | 'teacher' | 'all'
+  ) => {
     if (!user) {
       return redirect('/auth');
     }
+    if (allowedUsers !== 'all' && allowedUsers !== user.role) {
+      return redirect('/auth');
```

Comment:
Let's  throw an Error instead. We'll later create an error component to tell them they can't access that page, and offer to redirect them to the home page